File: backend/src/services/livroService.js
const prisma = require('../lib/prisma')
const openLibraryService = require('./openLibraryService')

async function buscar(termo, categoria, ordenar, pagina, tamanhoPagina, qualidadeMinima) {
  return openLibraryService.buscar(termo, categoria, ordenar, pagina, tamanhoPagina, qualidadeMinima)
}

async function tendencias() {
  return openLibraryService.tendencias()
}

async function detalhar(identificadorExterno) {
  return openLibraryService.detalhar(identificadorExterno)
}

async function obterOuCriarLivroLocal(identificadorExterno) {
  const existente = await prisma.livro.findUnique({ where: { identificadorExterno } })
  if (existente) return existente

  const detalhe = await openLibraryService.detalhar(identificadorExterno)
  return prisma.livro.create({
    data: {
      identificadorExterno: detalhe.identificadorExterno,
      titulo: detalhe.titulo,
      autor: detalhe.autor,
      descricao: detalhe.descricao,
      capa: detalhe.capa,
      editora: detalhe.editora,
      dataPublicacao: detalhe.dataPublicacao,
      categoria: detalhe.categoria,
      numeroPaginas: detalhe.numeroPaginas,
    },
  })
}

module.exports = { buscar, detalhar, tendencias, obterOuCriarLivroLocal }

File: backend/src/main/java/com/alexandria/service/LivroService.java
package com.alexandria.service;

import com.alexandria.dto.livro.BuscaLivrosResponse;
import com.alexandria.dto.livro.LivroDetalheResponse;
import com.alexandria.model.Livro;
import com.alexandria.repository.LivroRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class LivroService {

    private final GoogleBooksService googleBooksService;
    private final LivroRepository livroRepository;

    public LivroService(GoogleBooksService googleBooksService, LivroRepository livroRepository) {
        this.googleBooksService = googleBooksService;
        this.livroRepository = livroRepository;
    }

    public BuscaLivrosResponse buscar(String termo, String categoria, String ordenar,
                                       int pagina, int tamanhoPagina, boolean qualidadeMinima) {
        return googleBooksService.buscar(termo, categoria, ordenar, pagina, tamanhoPagina, qualidadeMinima);
    }

    public LivroDetalheResponse detalhar(String googleId) {
        return googleBooksService.detalhar(googleId);
    }

    @Transactional
    public Livro obterOuCriarLivroLocal(String googleId) {
        return livroRepository.findByIdentificadorExterno(googleId)
                .orElseGet(() -> {
                    LivroDetalheResponse detalhe = googleBooksService.detalhar(googleId);
                    Livro livro = Livro.builder()
                            .identificadorExterno(detalhe.identificadorExterno())
                            .titulo(detalhe.titulo())
                            .autor(detalhe.autor())
                            .descricao(detalhe.descricao())
                            .capa(detalhe.capa())
                            .editora(detalhe.editora())
                            .dataPublicacao(detalhe.dataPublicacao())
                            .categoria(detalhe.categoria())
                            .numeroPaginas(detalhe.numeroPaginas())
                            .build();
                    return livroRepository.save(livro);
                });
    }
}
